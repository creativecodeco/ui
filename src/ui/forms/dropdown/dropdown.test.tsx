import { fireEvent, render, screen } from '@testing-library/react';

import Dropdown from './dropdown.component';

import type { DropdownOption } from '@/types';

const options = Array.from({ length: 5 }).map<DropdownOption>((_, index) => ({
  value: index,
  label: `Option ${index}`
}));

describe('<Dropdown />', () => {
  it('snapshot', () => {
    const { container } = render(<Dropdown name='test' />);

    expect(container).toMatchSnapshot();
  });

  it('snapshot full', () => {
    const { container } = render(
      <Dropdown name='test' label='Label' isError error='Error' disabled />
    );

    expect(container).toMatchSnapshot();
  });

  it('default value', () => {
    const { getByTestId } = render(
      <Dropdown name='test' options={options} value='2' />
    );

    const control = getByTestId('test');

    expect((control as HTMLInputElement).value).toEqual('Option 2');
  });

  it('focus', () => {
    const { getByTestId } = render(<Dropdown name='test' options={options} />);

    const textBox = getByTestId('test');

    fireEvent.focus(textBox);

    expect(screen.getByTestId('options-test')).toBeInTheDocument();
  });

  it('focus disabled', async () => {
    const { getByTestId } = render(
      <Dropdown name='test' options={options} disabled />
    );

    const textBox = getByTestId('test');

    fireEvent.focus(textBox);

    const notExists = screen.queryByTestId('options-test');
    expect(notExists).not.toBeInTheDocument();
  });

  it('selects an option on click', () => {
    const onChangeMock = jest.fn();

    const { getByTestId, getByText } = render(
      <Dropdown name='test' options={options} onChange={onChangeMock} />
    );

    const textBox = getByTestId('test');
    fireEvent.focus(textBox);

    const optionToSelect = getByText('Option 1');
    fireEvent.click(optionToSelect);

    expect(onChangeMock).toHaveBeenCalledWith({
      target: { value: '1' }
    });
  });

  it('closes the dropdown on click outside', () => {
    const { getByTestId } = render(
      <Dropdown name='test' data-testid='textbox' options={options} />
    );

    const textBox = getByTestId('textbox');

    fireEvent.focus(textBox);

    fireEvent.change(textBox, { target: { value: 'NuevoValor' } });

    expect(screen.getByTestId('options-test')).toBeInTheDocument();

    fireEvent(
      document.body,
      new MouseEvent('mousedown', {
        bubbles: true,
        cancelable: true
      })
    );

    const notExists = screen.queryByTestId('options-test');
    expect(notExists).not.toBeInTheDocument();
  });

  it('handles change correctly', () => {
    const onTextChangeMock = jest.fn();
    const { getByTestId } = render(
      <Dropdown name='test' onTextChange={onTextChangeMock} options={options} />
    );
    const textBox = getByTestId('test');

    fireEvent.change(textBox, { target: { value: 'NuevoValor' } });

    expect(onTextChangeMock).toHaveBeenCalledWith('NuevoValor');
  });

  it('handles change correctly with option', () => {
    const onTextChangeMock = jest.fn();
    const { getByTestId } = render(
      <Dropdown name='test' onTextChange={onTextChangeMock} options={options} />
    );
    const textBox = getByTestId('test');

    fireEvent.change(textBox, { target: { value: 'Option' } });

    expect(onTextChangeMock).toHaveBeenCalledWith('Option');
  });
});
