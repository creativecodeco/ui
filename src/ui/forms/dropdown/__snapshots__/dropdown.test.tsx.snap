// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> snapshot 1`] = `
<div>
  <div
    class="dropdown block"
  >
    <div
      class="form-control w-full flex"
    >
      <div
        class="relative"
      >
        <input
          class="input input-bordered w-full text-box-size-md text-box-with-right-icon"
          data-testid="test"
          id="test"
          name="test"
          tabindex="0"
          value=""
        />
        <svg
          class="text-box-right-icon text-box-right-icon-size-md cursor-pointer"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 320 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Dropdown /> snapshot full 1`] = `
<div>
  <div
    class="dropdown block"
  >
    <div
      class="form-control w-full flex"
    >
      <div
        class="label"
      >
        <label
          class="label-text"
          for="test"
        >
          Label
        </label>
      </div>
      <div
        class="relative"
      >
        <input
          class="input input-bordered w-full text-box-size-md input-error text-box-with-right-icon"
          data-testid="test"
          disabled=""
          id="test"
          name="test"
          tabindex="0"
          value=""
        />
        <svg
          class="text-box-right-icon text-box-right-icon-size-md cursor-pointer"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 320 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41z"
          />
        </svg>
      </div>
      <p
        class="text-red-500"
      >
        Error
      </p>
    </div>
  </div>
</div>
`;
